pkg/upgrade/upgrade.go

```diff
@@ -217,6 +176,51 @@ func (u Upgrader) dryRunBinaries(upgradeBinaries []okctlUpgradeBinary, binaryPro
 	return nil
 }
 
+func (u Upgrader) doRunBinaries(upgradeBinaries []okctlUpgradeBinary, binaryProvider upgradeBinaryProvider) error {
+	for _, binary := range upgradeBinaries {
+		// Get
+		binaryRunner, err := binaryProvider.okctlUpgradeRunner(binary.RawVersion())
+		if err != nil {
+			return fmt.Errorf("getting okctl upgrade binary: %w", err)
+		}
+
+		_, _ = fmt.Fprintf(u.out, "--- Running upgrade: %s ---\n", binary)
+
+		// Run
+		_, err = binaryRunner.Run(okctlupgrade.Flags{
+			Debug:   u.debug,
+			Confirm: u.autoConfirm,
+		})
+
+		if err != nil {
+			_, _ = fmt.Fprintf(u.out, "--- Upgrade failed: %s ---\n", binary)
+			return fmt.Errorf("running upgrade binary %s: %w", binary, err)
+		}
+
+		// Mark as run
+		err = u.markAsRun(binary)
+		if err != nil {
+			return fmt.Errorf("marking upgrades as run: %w", err)
+		}
+
+		// Update cluster version
+		// Note that we don't save the hotfix version, only the semver version. This is because we use the cluster
+		// versioner later to validate that the current okctl version against the cluster version. The okctl version
+		// will always be a semver version without any hotfix in it, so it wouldn't be possible to store a hotfix
+		// version here.
+		//
+		// Also, a hotfix version is supposed to just fix any errors being made in an upgrade binary, so
```

Comment:
Nice to know. This means that we can only create one upgrade per okctl release. This should be communicated somewhere else as well. Maybe a short RFC

Reply:
Agh, that was not the intention. We ideally want to be able to have as many upgrades as we want for a release, for nice separation.

I'll have to look into it.

Reply:
As discussed in Slack, we do support multiple upgrades for one okctl version, because we support hotfixes.